include/dxc/Support/WinAdapter.h

```diff
@@ -563,27 +555,73 @@ enum tagSTATFLAG {
+
+constexpr uint8_t nybble_from_hex(char c) {
+  return ((c >= '0' && c <= '9')
+              ? (c - '0')
+              : ((c >= 'a' && c <= 'f')
+                     ? (c - 'a' + 10)
+                     : ((c >= 'A' && c <= 'F') ? (c - 'A' + 10)
+                                               : /* Should be an error */ -1)));
+  // if constexpr (c >= '0' && c <= '9')
+  //   return c - '0';
+  // else if constexpr (c >= 'a' && c <= 'f')
+  //   return c - 'a' + 10;
+  // else if constexpr (c >= 'A' && c <= 'F')
+  //   return c - 'A' + 10;
+  // throw -1;
+  // return -1;
+}
+
+constexpr uint8_t byte_from_hex(char c1, char c2) {
+  return nybble_from_hex(c1) << 4 | nybble_from_hex(c2);
+}
+
+constexpr uint8_t byte_from_hexstr(const char str[2]) {
+  return nybble_from_hex(str[0]) << 4 | nybble_from_hex(str[1]);
+}
+
+constexpr GUID guid_from_string(const char str[37]) {
+  // error: non-constant condition for static assertion :(
+  // static_assert(str[8] == '-', "Expected -");
+  // static_assert(str[13] == '-', "Expected -");
+  // static_assert(str[18] == '-', "Expected -");
+  // static_assert(str[23] == '-', "Expected -");
+  return GUID{static_cast<uint32_t>(byte_from_hexstr(str)) << 24 |
+                  static_cast<uint32_t>(byte_from_hexstr(str + 2)) << 16 |
+                  static_cast<uint32_t>(byte_from_hexstr(str + 4)) << 8 |
+                  byte_from_hexstr(str + 6),
+              static_cast<uint16_t>(
+                  static_cast<uint16_t>(byte_from_hexstr(str + 9)) << 8 |
+                  byte_from_hexstr(str + 11)),
+              static_cast<uint16_t>(
+                  static_cast<uint16_t>(byte_from_hexstr(str + 14)) << 8 |
+                  byte_from_hexstr(str + 16)),
+              {byte_from_hexstr(str + 19), byte_from_hexstr(str + 21),
+               byte_from_hexstr(str + 24), byte_from_hexstr(str + 26),
+               byte_from_hexstr(str + 28), byte_from_hexstr(str + 30),
+               byte_from_hexstr(str + 32), byte_from_hexstr(str + 34)}};
+}
+
+template <typename interface> inline GUID __emulated_uuidof();
+
+#define CROSS_PLATFORM_UUIDOF(interface, spec)                                 \
+  struct interface;                                                            \
+  template <> inline GUID __emulated_uuidof<interface>() {                     \
+    static const IID _IID = guid_from_string(spec);                            \
+    return _IID;                                                               \
+  }
+
+#define __uuidof(T) __emulated_uuidof<typename std::decay<T>::type>()
+
 #define IID_PPV_ARGS(ppType)                                                   \
-  (**(ppType)).uuidof(), reinterpret_cast<void **>(ppType)
+  __uuidof(decltype(**(ppType))), reinterpret_cast<void **>(ppType)
 
 #else // __EMULATE_UUID
 
-#define DECLARE_CROSS_PLATFORM_UUIDOF(T)
-#define DEFINE_CROSS_PLATFORM_UUIDOF(T)
+#define CROSS_PLATFORM_UUIDOF(interface, spec)                                 \
```

Comment:
Tex suggested an `#ifndef CROSS_PLATFORM_UUIDOF ` protection around this define. I didn't see any response to it. It seems a reasonable request to me.

Reply:
> I didn't see any response to it

Tex specifically requested to add it to the `dxcapi` variant, though he might have meant `WinAdapter` instead. That's indeed reasonable, and perhaps even required:

You requested to move this macro (in `dxcapi.h`) back up to where `DECLARE_CROSS_PLATFORM_UUIDOF` was previously defined:

```diff
-#ifdef _WIN32
-#define DECLARE_CROSS_PLATFORM_UUIDOF(T)
-#define DEFINE_CROSS_PLATFORM_UUIDOF(T)
-#else
+#ifndef _WIN32
 #include <dlfcn.h>
 #include "dxc/Support/WinAdapter.h"
 #endif
```

I'll make both changes :ok_hand: 

Note that I won't add protection around the `__EMULATE_UUID` variant: if `CROSS_PLATFORM_UUIDOF` gets defined prior it's probably the Windows (+msvc extensions) -only variant, an error that must be caught.

Reply:
Perhaps the comment references an old version of the code. It looks like you did a force push at some point so I can't go back and see what state the code was in when Tex made the comment. I took it to mean adding it to WinAdapter.h too, but that was in the context of the current state of the code. Maybe he was just asking for the ifndef that is now there. I really can't tell because the history was lost.

Reply:
I don't understand where it went either: GitHub used to be pretty good at showing "outdated" changes on the right line, even if commits are amended or force-pushed (it'll simply record the exact commit hash the review comment applies to). It must have been `WinAdapter.h`.